tutorials/3_transactions.md

```diff
@@ -99,127 +148,139 @@ struct Fund {
     sol_bool_t user_isOpen;
     sol_address_t user_owner;
 };
+
 struct Manager {
     sol_address_t model_address;
     sol_uint256_t model_balance;
     struct Fund user_fund;
 };
 ```
 
-Members prefixed by `user_` are taken from the smart contract definition while
-methods prefixed `model_` are generated by SmartACE. As we can see here, all
-primitive Solidity types are wrapped by singleton structures. To get the balance
-of `Fund` we must access `model_balance.v`.
+A c-struct is generated for each contract. All Solidity primitive types are
+wrapped by singleton c-struts in `primitives.h`. Members prefixed by `user_`
+are taken from the smart contract definition, while members prefixed by `model_`
+are generated by SmartACE. If a contract is constructed with `new`, it is nested
+inside the contract which calls `new`. To illustrate, if we defined a contract
```

Comment:
I don't quite understand the nesting for dynamically constructed contracts. Why isn't the type of `user_fund` a pointer to the struct you allocate when constructing the new fund?

Reply:
Is this implicitely encoding some sort of ownership?

Reply:
There is an explicit assumption that's we're encoding a fixed number of smart contracts.

There's an implicit assumption (presently) that the number is fixed by restricting `new` to construction, and by requiring a 1-to-1 mapping between state variables and `new` calls.

This could be taken as ownership.

I've thought about relaxing some of the implicit assumptions but we haven't hit an example where it mattered.

I can elaborate on this in the post.

Reply:
I see. I wasn't aware of the assumption about encoding a fixed number of contracts. Does the tool enforce it by rejecting programs that allocate new contracts outside constructors?

I wonder if we can relax the assumption similar to how we use interference for user addresses.

I wouldn't elaborate on it in the post. In fact, I'd probably be more vague about this since it isn't crucial for understanding the post.

Reply:
That's essentially what we do.

I have had a few ideas on how to relax this, driven by examples we've come across seen.

1. Some smart contracts go through stages, and each stage is associated with a newly constructed helper contract. The number of stages in these contracts are usually bounded, so we could still do this case statically.
2. Some contracts take in addresses, and then cast those addresses to contracts (I believe this is what you're referencing). In this case, I had the same idea you were suggesting.

I'll reword it to imply "in this specific case, this is what we do."

Reply:
I looked at the wording again. It should have said "If a state variable of the contract type is initialized once by a call to `new`, ...". I think it's okay now.

Reply:
Sounds good. Yeah, (2) is definitely quite common (e.g., for tokens that are used by a contract). I was thinking that even for a dynamically created contract (with unbounded number of creations) one could use an "interference contract" to summarize the behavior.